agent/src/main/scala/za/co/absa/atum/agent/model/MeasureResult.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2021 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.atum.agent.model
+
+import za.co.absa.atum.agent.exception.AtumAgentException.MeasurementException
+import za.co.absa.atum.model.dto.MeasureResultDTO.ResultValueType
+
+/**
+ *  This trait defines a contract for a measure result.
+ */
+trait MeasureResult {
+  val resultValue: Any
```

Comment:
Isn't better to use generic `T` instead of `Any`? If we were to use generic `T` we would preserve type safety (`Any` can be any type at all) and flexibility (you cannot apply constraints like contexts bounds to `Any`).

Reply:
Yes I'm absolutely aware of that. The thing is that if this would be of a generic type, then this trait would need to be of a generic type as well, and every usages of it would need to take that into an account. I gave it a try (1-2 hours) twice in the past and I didn't like the solution at all. I even wanted to include `Shapeless` here and this would be a the first step, but as I said I decided to keep it like this for now

Reply:
@TebaleloS @benedeki 
I am interested in opinion of others on this matter.

Reply:
@lsulak 
Can you maybe elaborate a bit more on the problems the generic would bring? I understand that the type parameter will be reflected in signatures throughout the code but when you create an instance of a class you don't need to specify the type as it's inferred.

Reply:
```scala
trait MeasureResult[T] {
  val resultValue: T
  val resultValueType: ResultValueType.ResultValueType
}
```

basically this ^ would mean, that I would need to adjust code in `MeasurementProcessor` and its usecases, `Measurement` and its usecases, `MeasurementBuilder` and its usecases, including things like `AtumContext` -> `takeMeasurements` and then also unit tests and god knows what else :D 

I really like David's solution though, the whole 'generic' problem is nicely 'encapsulated' within the whole module

Reply:
I think you are actually right about this one. Anyway I have one suggestion, to replace `Enumeration` with `ADT` as shown below. Using sealed traits and case classes for `ResultValueType` is more idiomatic Scala and provides more flexibility. In this setup, `ResultValueType` is a sealed trait, which means it can only be extended in the same file. `StringResult`, `LongResult`, `BigDecimalResult` and `DoubleResult` are case classes the extend `ResultValueType`. This limits the possible types of `ResultValueType` to these four case classes.

This approach provides more flexibility because you can add methods and fields to the case classes if needed. It also provides more type safety because the compiler can check that you've handled all possible cases when you use pattern matching.
 
```scala
sealed trait ResultValueType
case class StringResult(value: String) extends ResultValueType
case class LongResult(value: Long) extends ResultValueType
case class BigDecimalResult(value: BigDecimal) extends ResultValueType
case class DoubleResult(value: Double) extends ResultValueType

trait MeasureResult {
  val resultValue: ResultValueType
}
```

Reply:
I don't know. I know that this is a common Scala pattern, but the original solution with the Enumerate is extremely small, really easy to refactor and extend if needed:

```scala
  object ResultValueType extends Enumeration {
    type ResultValueType = Value
    val String, Long, BigDecimal, Double = Value
```

Reply:
I leave it up to you to consider these options and decide. My personal preference would be the idiomatic solution with sealed trait and case classes. The reasoning was mentioned above, plus you wouldn't have to have a separate field just to keep track of the type.